utils_cv/tracking/model.py

```diff
@@ -0,0 +1,240 @@
+        self.opt.arch = arch
+        self.opt.head_conv = head_conv
+        self.opt.down_ratio = down_ratio
+
+        # TODO setup logging
+
+    def fit(
+        self,
+        data_root,  # consider making a custom cvbp Dataset
+        data_path,
+        im_size,
+        lr: float = 1e-4,
+        lr_step: str = "20,27",
+        num_epochs: int = 30,
+        batch_size: int = 12,
+        num_iters: int = -1,
+        val_intervals: int = 5,
+        num_workers: int = 8,
+    ) -> None:
+        self.opt.lr = lr
+        self.opt.lr_step = lr_step
+        self.opt.num_epochs = num_epochs
+        self.opt.batch_size = batch_size
+        self.opt.num_iters = num_iters
+        self.opt.val_intervals = val_intervals
+        self.opt.num_workers = num_workers
+        self.opt.device = torch_device()
+
+        # initialize dataset
+        dataset = self._init_dataset(data_root, train_path, im_size) # replaces the step of reading data_cfg
+        self.opt.update_dataset_info_and_set_heads(dataset)
+
+        train_loader = DataLoader(
+            dataset,
+            batch_size=self.opt.batch_size,
+            shuffle=True,
+            num_workers=self.opt.num_workers,
+            pin_memory=True,
+            drop_last=True,
+        )
+
+        # initialize model
+        model = create_model(self.opt.arch, self.opt.heads, self.opt.head_conv)
+        optimizer = torch.optim.Adam(model.parameters(), self.opt.lr)
+        start_epoch = 0
+        if self.opt.load_model != "":
+            model, optimizer, start_epoch = load_model(
+                model,
+                self.opt.load_model,
+                optimizer,
+                self.opt.resume,
+                self.opt.lr,
+                self.opt.lr_step,
+            )
+
+        Trainer = train_factory[self.opt.task]
+        trainer = Trainer(self.opt, self.model, optimizer)
+        trainer.set_device(
+            self.opt.gpus, self.opt.chunk_sizes, self.opt.device
+        )
+        best = 1e10
+
+        # training loop
+        for epoch in range(start_epoch + 1, self.opt.num_epochs + 1):
+            mark = epoch if self.opt.save_all else "last"
+            log_dict_train, _ = trainer.train(epoch, train_loader)
+            ## TODO logging
+            if (
+                self.opt.val_intervals > 0
+                and epoch % self.opt.val_intervals == 0
+            ):
+                save_model(
+                    osp.join(self.opt.save_dir, f"model_{mark}.pth"),
+                    epoch,
+                    model,
+                    optimizer,
+                )
+            else:
+                save_model(
+                    osp.join(
+                        self.opt.save_dir,
+                        f"model_last.pth",
+                        epoch,
+                        model,
+                        optimizer,
+                    )
+                )
+
+        for epoch in self.opt.lr_step:
+            save_model(
```

Comment:
[minor] Let's maybe totally remove all model saving inside fit() but show in the notebook how to save the model manually.